[Step3] 로또(자동) PR 입니다

songsimo opened the pull request:
안녕하세요, 좋은 하루입니다, 리뷰어님 😊

3번째 PR 요청입니다.

마지막에 로또(수동) 기입이 있어서 로또 생성을 추상화해보려다가 잘 안 됐네요.

"*Validator" 클래스나 "*Utils" 클래스를 안 만드려고(엘레강트 오브젝트 참고) 하다보니 소스가 더 더러워진건 아닌가라는

생각도 들었어요. 

그리고 한가지 과제와는 관련 없는 질문이 있는데요,  고민이 들어서요, 죄송합니다 😅😅.

엘레강트 오브젝트의 내용을 적용해보려고 하고 있어요. 과제를 진행하면서 생성자를 많이 만들려고 하고 있습니다.

```
궁금한 것은 주 생성자 하나와 부 생성자  여러 개를 만들라는데 생성자에 코드를 넣지 말라라고 하면, 검증이나 파싱은

어떻게 처리하라는 걸까요? 참고할 만한 것을 못 찾고 있어요😳😳...
```

시간 내주셔서 감사합니다, 리뷰 잘 부탁드려요!

songsimo committed:
fix: 피드백 반영

- 테스트 코드 @NullAndEmptySource 적용
- 총합 구하는 메소드명 변경(의미 명확하게 전달)
- 유효한 숫자 검증 예외처리 변경(RuntimeException -> IllegalArgumentException)
- Pattern 상수화 미진한 부분 추가 수정

songsimo committed:
docs: 요구사항 작성(README.md)

songsimo committed:
feat: 로또 번호 범위 1-45 검증 기능 구현

songsimo committed:
feat: 로또 번호 개수 검증

- 로또 번호 개수는 6개

songsimo committed:
fix: 로또 번호 범위 검증 수정

- MAX 값 잘못 기입하여 수정(1->45)

songsimo committed:
feat: 번호 일치 개수 기능 구현

songsimo committed:
docs: 숫자 중복 기입 검증 추가

songsimo committed:
feat: 중복된 숫자 입력 검증

songsimo committed:
feat: 로또 번호를 자동 생성 기능

songsimo committed:
feat: InputView 추가

songsimo committed:
feat: 금액 입력은 0 이상 검증

songsimo committed:
feat: 금액의 null 혹은 빈값 검증

songsimo committed:
feat: 금액 입력 숫자만 가능 검증

songsimo committed:
feat: 문자열 리스트로 Lotto 생성 기능 추가

songsimo committed:
feat: LottoMachine 추상화 추가

songsimo committed:
docs: 불필요하게 기입한 요구사항 삭제

songsimo committed:
feat: 로또 결과 얻기 기능 추가

- Rank(Enum) 추가하여 당첨 결과 얻기
- RankResult 추가: 결과 통계 얻기
- DTO 추가: 상금 정보, 당첨 개수에 대한 DTO

songsimo committed:
feat: OutputView 추가

- 구입 개수 출력
- 로또 번호 출력

songsimo committed:
fix: 로또 사이즈 검증 수정

songsimo committed:
feat: 수익률 계산 기능 추가

songsimo committed:
feat: LottoController 추가

songsimo commented:
https://github.com/next-step/java-lotto-pro/pull/418#issuecomment-1126244157

이어서 궁금한 사항이 있어서 여기에 코멘트  달아요. 말씀하신 내용 중에

```
힌트 5는 커스텀 구분자 (:= 다른 요구사항)에 대한 요구사항인 만큼 커밋을 나누어 볼 것 같습니다 :)
```

내용은 힌트 5는 비슷하지만 다른 요구사항이라고 보신건가요? 그렇다면 어떠한 점에서 다른 요구사항이라 보신 건가요?🙂🙂

저는 ```문자열을 나눈다``` 라는 기능으로 본다면 힌트 3,4,5는 같은 커밋 단위로 묶여도 되지 않을까? 라는 생각을 했어요.

코멘트 달아주신 것처럼 커밋내 코드량이 300 ~ 400 줄이 넘어가면 커밋을 나누는 점... 좋은 깨달음입니다, 감사합니다 = )

enemfk777 commented on src/main/java/lotto/domain/AbstractLottoMachine.java:
```diff
@@ -0,0 +1,5 @@
+
+public abstract class AbstractLottoMachine {
+	public static final int LOTTO_PRICE = 1000;
+}
```
이 클래스의 책임은 무엇으로 설정하셨을까요?
추상클래스이지만 멤버 변수나 추상메서드가 없어
마커인터페이스 정도의 기능을 하고 있는 것 같습니다.

추상클래스와 인터페이스에 대해 다시 한 번 고민해 보시면 좋을 것 같아요 :)

또 LOTTO_PRICE의 위치가 어색해 보이는데요,
Money 클래스의 인스턴스가 어플리케이션 내에서 의미를 가지려면 최소 1000원 이상의 값은 가져야 할 것 같은데,
그렇다면 LOTTO_PRICE가 높은 응집도를 가지려면 어느 클래스에 위치해야 할까요? :)

enemfk777 commented on src/main/java/lotto/domain/AutoLottoMachine.java:
```diff
@@ -0,0 +1,24 @@
+		numbers = IntStream.range(Number.MIN_NUMBER, Number.MAX_NUMBER)
+						.mapToObj(Number::new)
+						.collect(Collectors.toList());
+	}
```
스트림 API를 활용한 숫자 풀 캐싱 좋습니다 👍

식 한 줄로 끝나니 static 블럭 없이 바로 할당해도 될 것 같아요 :)

enemfk777 commented on src/main/java/lotto/domain/Numbers.java:
```diff
@@ -0,0 +1,56 @@
+public class Numbers {
+	private final String VALID_DUPLICATION = "중복된 숫자가 존재합니다.";
+
+	private List<Number> numbers;
```
List 외에 중복을 허용하지 않는 자료구조로 어떤게 있을까요? :)

enemfk777 commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.stream.Stream;
+
+public enum Rank {
+	FIRST(6, 2000000000),
```
2_000_000_000 와같이 언더바를 넣어주면 코드의 가독성을 높일 수 있습니다 :)

enemfk777 commented on src/main/java/lotto/domain/RankResult.java:
```diff
@@ -0,0 +1,59 @@
+import lotto.dto.PrizeReport;
+
+public class RankResult {
+	private int firstCount = 0;
+	private int secondCount = 0;
+	private int thirdCount = 0;
+	private int fourthCount = 0;
+
+	public void setUp(Rank rank) {
+		if(rank.isFirst()) {
+			firstCount += 1;
+			return;
+		}
+
+		if(rank.isSecond()) {
+			secondCount += 1;
+			return;
+		}
+
+		if(rank.isThird()) {
+			thirdCount += 1;
+			return;
+		}
+
+		if(rank.isFourth()) {
+			fourthCount += 1;
+			return;
+		}
+	}
```
Map을 사용하면 좀더 깔끔한 코드를 만들 수 있을 것 같아요 :)

songsimo replied:
오!!!!!!! 맞아요 맞아요!! 감사합니다, 리뷰어님 = ) 훨씬 깔끔하게 작성되었네요 🙇🙇

enemfk777 commented on src/main/java/lotto/dto/LottoNumber.java:
```diff
@@ -0,0 +1,24 @@
+	public List<Integer> getNumbers() {
+		return numbers;
+	}
+}
```
Numbers 와 이 클래스는 역할이 겹치는 것 같은데 어떻게 생각하시나요?

enemfk777 commented on src/main/java/lotto/view/InputView.java:
```diff
@@ -0,0 +1,20 @@
+	private static final String INPUT_MONEY_MESSAGE = "구입금액을 입력해 주세요.";
+	private static final String INPUT_WINNING_NUMBER_MESSAGE = "지난 주 당첨 번호를 입력해 주세요.";
+
+	private static final Scanner scanner = new Scanner(System.in);
```
스캐너 캐싱 👍

enemfk777 commented on src/main/java/lotto/view/InputView.java:
```diff
@@ -0,0 +1,20 @@
+
+	public static String inputMoney() {
+		System.out.println(INPUT_MONEY_MESSAGE);
+		return scanner.next();
```
금액은 nextInt()로 바로 정수형을 받는건 어떨까요?
그러면 입력값의 타입 (숫자인지 문자인지) validation도 view에서 책임질 수 있을것같아요 :)

songsimo replied:
아! 그러네요. 왜 문자열로만 받으려고 했는지... ;;; 감사합니다

enemfk777 commented on src/main/java/lotto/view/OutputView.java:
```diff
@@ -0,0 +1,41 @@
+import lotto.dto.LottoNumber;
+import lotto.dto.PrizeReport;
+
+public class OutputView {
```
InputView 처럼 문자열 상수화를 활용해보시면 어떨까요? :)

enemfk777 commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,62 @@
+		}
+
+		return lottoTickets;
+	}
```
로또 생성책임이 Controller의 책임이라고 볼수 있을까요? :)
실무에서 프론트엔드에 제공하는 API의 컨트롤러를 빗대어 생각해보시면 
직관적으로 이해가 되실 것 같습니다 ! :)

로또 생성책임을 갖는 클래스를 분리해보시는건 어떨까요?

enemfk777 commented on src/main/java/lotto/domain/LottoTickets.java:
```diff
@@ -0,0 +1,41 @@
+
+	public void addLotto(Lotto lotto) {
+		lottoTickets.add(lotto);
+	}
```
객체 외부에서 내부의 컬렉션을 조작할 수 있는 로직은 최대한 지양하는 편이 좋습니다.
객체 내부의 상태를 추측하기 어려워지기 때문인데요
(무조건 안된다는 것은 아닙니다. 로직에 따라 필요할 때도 있곤 하죠 :))

LottoTickets 같은 경우 addLotto가 현재 처음 초기화할때만 쓰이고 있으므로
로직상 반드시 필요한 경우는 아니라고 생각됩니다.

그렇다면 그냥 생성자로만 초기화하시는건 어떨까요?

enemfk777 commented:
> 궁금한 것은 주 생성자 하나와 부 생성자  여러 개를 만들라는데 생성자에 코드를 넣지 말라라고 하면, 검증이나 파싱은
> 
> 어떻게 처리하라는 걸까요? 참고할 만한 것을 못 찾고 있어요😳😳...

가장 간단한 방법은 정적팩토리메서드를 사용하시는 방법입니다.
생성자에서는 인자로 받은 값의 할당만 진행하고
validation 등은 정적팩토리 메서드내에서 수행하는 것이죠

두 번째 방법은 생성자에서 다양한 타입의 인자 받아 초기화 하고 싶은 멤버 변수의 타입을 인터페이스로 설정 한 후
인자마다 각각 구현체로 만들어 초기화 하는 방법입니다.
(참고 : https://jackjeong.tistory.com/149?category=883547)

그리고 생성자를 여러개 만들기전에 먼저 생각해보아야 할 것은
**여러개의 생성자가 현재 로직에서 정말로 필요한가?** 를 먼저 살펴보는 것입니다.

짧고 간결한 코드일수록 읽기 쉽고 유지보수성도 좋을수 밖에 없습니다 :)
생성자가 하나 늘어날 수록 무엇을 위해 존재하는 생성자인지 파악해야하는 부담도 가중 되게 됩니다.

예를 들면 어떤 어플리케이션에서 하나의 도메인 객체를 초기화 하는데 View v1으로부터  int를 인자로 받았다고 해보죠
그런데 View v2 가 나오면서 String을 인자로 넘겨주게 되는 상황이 되었다고 해볼게요.
그리고 View v2의 수정은 불가능한 상황이고 V1과 V2를 동시에 지원해야 하는 상황입니다.
이런 상황이라면 로직상 여러개의 생성자가 정말로 필요해진 경우라고 볼 수 있겠죠 :)

답변을 달다보니 코멘트가 좀 길어졌네요 :)
다시 한 번 강조하자면 중요한건 **생성자는 풍부해야한다**보다 **풍부한 생성자가 필요한가** 를 먼저 고민해보시면 좋을 것 같아요

enemfk777 reviewed (changes requested):
철현님 안녕하세요 :) 
3단계 미션 진행 잘해주셨네요 :)
몇가지 개선점에 대해 코멘트 달아 드렸으니 반영 후 다시 한 번 요청 부탁드립니다.!
또 질문 하신 사항에 대해선 PR 코멘트로 달아드렸으니 확인부탁드려요 :)

화이팅입니다 💪

songsimo committed:
fix: 피드백 반영

- 자료구조 변경: 숫자 중복 검증
- Rank(Enum) 상금 가독성 좋게 처리
- RankResult Map 처리로 가독성 좋게 처리

등...

enemfk777 commented on src/main/java/lotto/domain/LottoTickets.java:
```diff
@@ -0,0 +1,41 @@
+		return rankResult;
+	}
+
+	public void automaticallyGenerate() {
+		lottoTickets.add(AutoLottoNumberStrategy.generate());
+	}
```
개발자가 제어 할 수 없는 코드가 들어있는 코드는 테스트코드를 작성하기 어렵게하고,
내부 상태를 예측하기 어렵게 합니다.
아래 아티클을 읽어보시고, 개발자가 제어 할 수 없는 부분을 분리해 보시길 바래요 :)
https://tecoble.techcourse.co.kr/post/2020-05-07-appropriate_method_for_test_by_parameter/

enemfk777 commented on src/test/java/lotto/domain/LottoTest.java:
```diff
@@ -0,0 +1,33 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+public class LottoTest {
```
전체적으로 테스트코드가 부족하네요 :)

View클래스를 제외한 모든 클래스들의
public 메서드들에 대한 테스트코드를 작성해주세요 :)

테스트코드를 작성하실 때 중요한 사항중 하나는 테스트코드만을 위한 코드를 추가하면 안된다는 점입니다.
테스트코드는 프로덕션 코드를 테스트하기 위한 코드일뿐
테스트코드가 프로덕션 코드를 오염시켜서는 안됩니다. :)

위 원칙을 지키면서 테스트코드를 작성하시다보면 테스트가 불가능해보이는 퍼블릭 메서드도 존재 할 수 있는데요,
그럴 땐 클래스 분리가 잘못된건 아닌지 고민해보시면 좋을것 같아요 😄

enemfk777 reviewed (approved):
안녕하세요 철현님 :)

코멘트를 잘 반영해주셨습니다.
하지만 아직 테스트코드가 많이 부족하네요. 😭
다음 단계 진행하면서 View클래스를 제외한 클래스들 모두 테스트코드 작성 부탁드립니다..!
이번 단계는 이만 머지 하겠습니다
고생하셨습니다 :)
